src/app/clusters/thermostat-server/thermostat.cpp

```diff
@@ -45,115 +45,182 @@ void emberAfThermostatClusterServerInitCallback(void)
 }
 
-bool emberAfThermostatClusterSetWeeklyScheduleCallback(app::CommandHandler * commandObj, uint8_t numberOfTransitionsForSequence,
-                                                       uint8_t daysOfWeekForSequence, uint8_t modeForSequence, uint8_t * payload)
+bool emberAfThermostatClusterSetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t numberOfTransitionsForSequence, uint8_t daysOfWeekForSequence,
+                                                       uint8_t modeForSequence, uint8_t * payload)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterSetpointRaiseLowerCallback(app::CommandHandler * commandObj, uint8_t mode, int8_t amount)
+int32_t EnforceHeatingSetpointLimits(int32_t HeatingSetpoint, EndpointId endpoint)
+{
+    // Optional Mfg supplied limits
+    int16_t AbsMinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t AbsMaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Optional User supplied limits
+    int16_t MinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t MaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Attempt to read the setpoint limits
+    // Absmin/max are manufacturer limits
+    // min/max are user imposed min/max
+
+    // Note that the limits are initialized above per the spec limits
+    // if they are not present emberAfReadAttribute() will not update the value so the defaults are used
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMinHeatSetpointLimit, sizeof(AbsMinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMaxHeatSetpointLimit, sizeof(AbsMaxHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MinHeatSetpointLimit, sizeof(MinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MaxHeatSetpointLimit, sizeof(MaxHeatSetpointLimit), NULL);
+
+    // Make sure the user imposed limits are within the manufacture imposed limits
+
+    if (MinHeatSetpointLimit < AbsMinHeatSetpointLimit)
+        MinHeatSetpointLimit = AbsMinHeatSetpointLimit;
+
+    if (MaxHeatSetpointLimit > AbsMaxHeatSetpointLimit)
+        MaxHeatSetpointLimit = AbsMaxHeatSetpointLimit;
+
+    if (HeatingSetpoint < MinHeatSetpointLimit)
+        HeatingSetpoint = MinHeatSetpointLimit;
+
+    if (HeatingSetpoint > MaxHeatSetpointLimit)
+        HeatingSetpoint = MaxHeatSetpointLimit;
```

Comment:
Where is this clamping behavior specified in the spec?  The spec says (for `MinHeatSetpointLimit`, say):

> The MinHeatSetpointLimit attribute specifies the minimum level that the heating setpoint MAY be set to

which to me implies that it's not allowed to be set to a level below this, and attempts to do so should lead to errors.  But it's not very clear, and perhaps the spec here needs some clarification?

Reply:
As you point out the spec is not clear.  I do not believe that the clamping behavior is against the spec.

Reply:
My point is that we need a spec issue filed to track the spec being unclear.  That issue need not block this PR, but it needs to exist.  And we should probably have a code comment referencing it...

Reply:
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3725